### PR TITLE
support schedule based on cluster provider/region/zone

### PR DESCRIPTION
**Pull request by mrlihanbo**
Signed-off-by: lihanbo <lihanbo2@huawei.com>

**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:
support schedule based on cluster provider/region/zone

**Which issue(s) this PR fixes**:
Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:
"NONE"

### CODE REVIEW

File: pkg/webhook/clusterpropagationpolicy/validating.go [hunk 1]
```diff
@@ -50,6 +50,11 @@ func (v *ValidatingAdmission) Handle(ctx context.Context, req admission.Request)
 		return admission.Denied(err.Error())
 	}
 
+	if err := helper.ValidateFieldSelector(policy.Spec.Placement.ClusterAffinity.FieldSelector); err != nil {
+		klog.Info(err)
```
Comment (carmark) [inline pkg/webhook/clusterpropagationpolicy/validating.go:54]: `klog.Error(err)`  ?

File: pkg/webhook/propagationpolicy/validating.go [hunk 2]
```diff
@@ -50,6 +50,11 @@ func (v *ValidatingAdmission) Handle(ctx context.Context, req admission.Request)
 		return admission.Denied(err.Error())
 	}
 
+	if err := helper.ValidateFieldSelector(policy.Spec.Placement.ClusterAffinity.FieldSelector); err != nil {
+		klog.Info(err)
```
Comment (carmark) [inline pkg/webhook/propagationpolicy/validating.go:54]: same above.

File: pkg/util/selector.go [hunk 3]
```diff
@@ -46,7 +56,7 @@ func ResourceMatches(resource *unstructured.Unstructured, rs v1alpha1.ResourceSe
 }
 
 // ClusterMatches tells if specific cluster matches the affinity.
-// TODO(RainbowMango): Now support ClusterAffinity.ClusterNames, ClusterAffinity.ExcludeClusters and ClusterAffinity.LabelSelector. More rules will be implemented later.
+// Now support ClusterAffinity.ClusterNames, ClusterAffinity.ExcludeClusters , ClusterAffinity.LabelSelector and FieldSelector.
```
Comment (carmark) [inline pkg/util/selector.go:59]:
```
// Now support ClusterAffinity.ClusterNames, ClusterAffinity.ExcludeClusters, ClusterAffinity.LabelSelector and FieldSelector.
```
?

File: pkg/util/selector.go [hunk 4]
```diff
@@ -1,14 +1,24 @@
...
+	"k8s.io/apimachinery/pkg/selection"
 
 	clusterv1alpha1 "github.com/karmada-io/karmada/pkg/apis/cluster/v1alpha1"
 	"github.com/karmada-io/karmada/pkg/apis/policy/v1alpha1"
 )
 
+const (
+	providerField = "provider"
+	regionField   = "region"
+	zoneField     = "zone"
```
Comment (carmark) [inline pkg/util/selector.go:19]: make them as common variables? so the [case](https://github.com/karmada-io/karmada/pull/357/files#diff-4fe5d873ba71c3512cafdeeb480dce209710dd2b5ed84d900522e1bc07aa7f90R86) can use them.

**Comment by mrlihanbo** [comment]
@carmark Thanks for your review. I have added e2e test, PTAL, Thanks

File: pkg/util/helper/policy.go [hunk 5]
```diff
@@ -73,3 +75,26 @@ func IsDependentClusterOverridesPresent(c client.Client, policy *policyv1alpha1.
 
 	return true, nil
 }
+
+// ValidateFieldSelector tests if the fieldSelector is valid.
+func ValidateFieldSelector(fieldSelector *policyv1alpha1.FieldSelector) error {
```
Comment (RainbowMango) [inline pkg/util/helper/policy.go:80]: Move to `pkg/util/validation` package and might need to rename the function.

File: pkg/util/helper/policy.go [hunk 6]
```diff
@@ -73,3 +75,26 @@ func IsDependentClusterOverridesPresent(c client.Client, policy *policyv1alpha1.
...
+	if fieldSelector == nil {
+		return nil
+	}
+
+	for _, matchExpression := range fieldSelector.MatchExpressions {
+		switch matchExpression.Key {
+		case util.ProviderField, util.RegionField, util.ZoneField:
```
Comment (RainbowMango) [inline pkg/util/helper/policy.go:87]: Upcase or lowercase?

File: test/e2e/fieldselector_test.go [hunk 7]
```diff
@@ -0,0 +1,121 @@
...
+	"github.com/onsi/ginkgo"
+	"github.com/onsi/gomega"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/rand"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	policyv1alpha1 "github.com/karmada-io/karmada/pkg/apis/policy/v1alpha1"
+	"github.com/karmada-io/karmada/pkg/util"
+
```
Comment (RainbowMango) [inline test/e2e/fieldselector_test.go:16]: Remove empty line?

File: test/e2e/fieldselector_test.go [hunk 8]
```diff
@@ -0,0 +1,121 @@
...
+					}
+					fmt.Printf("setting provider and region for cluster %v", cluster)
+					clusterObj := &clusterv1alpha1.Cluster{}
+					err := controlPlaneClient.Get(context.TODO(), client.ObjectKey{Name: cluster}, clusterObj)
+					gomega.Expect(err).ShouldNot(gomega.HaveOccurred())
+
+					clusterObj.Spec.Provider = providerMap[index]
+					clusterObj.Spec.Region = regionMap[index]
+
+					err = controlPlaneClient.Update(context.TODO(), clusterObj)
```
Comment (RainbowMango) [inline test/e2e/fieldselector_test.go:74]: This will override the original settings and never get them back.

**Comment by RainbowMango** [comment]
@carmark Do you have any other comments?

**Comment by carmark** [comment]
@RainbowMango No others, thanks.

**Pull request merged**